### PR TITLE
Ensure desired_vcpus is set before including it in update

### LINKED ISSUE #4671: resource aws_batch_compute_environment gives error if optional desired_vcpus is not set
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
Terraform v0.11.7
+ provider.aws v1.19.0
```

### Affected Resource(s)

* aws_batch_compute_environment

### Terraform Configuration Files

```hcl

resource aws_batch_compute_environment batch_env {
  compute_environment_name = "postgres-loader"
  service_role             = "arn:aws:iam::XXXXXXXXXX:role/service-role/AWSBatchServiceRole"
  type                     = "MANAGED"

  compute_resources {
    type          = "EC2"
    instance_role = "arn:aws:iam::XXXXXXXXXX:instance-profile/ecsInstanceRole"
    instance_type = ["optimal"]
    min_vcpus     = 1
    max_vcpus     = 32

    subnets            = ["${local.subnet_ids}"]
    security_group_ids = ["${aws_security_group.batch.id}"]

    tags {
      Name        = "insights"
      Environment = "production"
    }
  }
}
```

### Debug Output

https://gist.github.com/jasantunes/397662245f1877848215e69182c1ecf4

### Panic Output

N/A.

### Expected Behavior

It should've accepted the `min_vcpu` and `max_vcpu` parameters and infer nothing about parameter `desired_vcpus` that was not set and is optional.

### Actual Behavior

```
Error: Error applying plan:

1 error(s) occurred:

* aws_batch_compute_environment.batch_env: 1 error(s) occurred:

* aws_batch_compute_environment.batch_env: : Error executing request, Exception : desiredvCpus should be between minvCpus and maxvCpus, RequestId: 06362c77-6133-11e8-8122-077fbc0c56df
	status code: 400, request id: 06362c77-6133-11e8-8122-077fbc0c56df
```

### Steps to Reproduce

Create a `aws_batch_compute_environment` resource without setting `desired_vcpus` parameter in `compute_resources`.

### Important Factoids

N/A.

### References

* https://docs.aws.amazon.com/batch/latest/userguide/compute_environment_parameters.html
* https://www.terraform.io/docs/providers/aws/r/batch_compute_environment.html
* https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_batch_compute_environment.go

### PR DESCRIPTION
**Pull request by hectcastro**
Fixes #4671 #4788 

When updating an AWS Batch compute environment, do not specify `DesiredvCpus` in the call to `UpdateComputeEnvironment` unless `desired_vcpus` is set.

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSBatchComputeEnvironment_createEc2'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccAWSBatchComputeEnvironment_createEc2 -timeout 120m
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccAWSBatchComputeEnvironment_createEc2
--- PASS: TestAccAWSBatchComputeEnvironment_createEc2 (49.55s)
=== RUN   TestAccAWSBatchComputeEnvironment_createEc2WithTags
--- PASS: TestAccAWSBatchComputeEnvironment_createEc2WithTags (46.84s)
=== RUN   TestAccAWSBatchComputeEnvironment_createEc2WithoutComputeResources
--- PASS: TestAccAWSBatchComputeEnvironment_createEc2WithoutComputeResources (21.39s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       117.838s
```

### CODE REVIEW

**Comment by bflad** [comment]
Hey @hectcastro! Can you add an acceptance test which tries to cover this behavior? I have a feeling the proposed fix should actually be adding a `DiffSuppressFunc` to the `desired_vcpus` attribute (returning true when the new value is 0) rather than trying to adjust the read/update logic. Please feel free to reach out if you have any questions.

**Comment by hectcastro** [comment]
I think I've got a decent test case for not setting `desired_vcpus` and setting `min_vcpus` to greater than zero:

```
$ make testacc TESTARGS='-run=TestAccAWSBatchComputeEnvironment_updateMinvCpus'               
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccAWSBatchComputeEnvironment_updateMinvCpus -timeout 120m                                                                                    
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccAWSBatchComputeEnvironment_updateMinvCpus
--- FAIL: TestAccAWSBatchComputeEnvironment_updateMinvCpus (63.36s)
        testing.go:518: Step 1 error: Error applying: 1 error(s) occurred:

                * aws_batch_compute_environment.ec2: 1 error(s) occurred:

                * aws_batch_compute_environment.ec2: : Error executing request, Exception : desiredvCpus should be between minvCpus and maxvCpus, RequestId: 7f8fdc3b-73b1-11e8-bbb8-d18d8e472e6e
                        status code: 400, request id: 7f8fdc3b-73b1-11e8-bbb8-d18d8e472e6e                                                                                       
FAIL
FAIL    github.com/terraform-providers/terraform-provider-aws/aws       63.403s
make: *** [testacc] Error 1
```

I had a hard time implementing a fix with `DiffSuppressFunc` (thanks for the documentation pointers—I didn't know they existed, and they were helpful). Even if I return `true` all of the time from that function, this error still occurs. I think that's because although it suppress the diff, it still passes along the value via the API call.

In trying to make a better sense of `desired_vcpus`, it behaves a bit like a computed value (AWS can change it out from under you in a `MANAGED` CE depending on jobs in the queue). From the Terraform configuration DSL, we've been able to suppress errors by looking up the current `desired_vcpu` count via the AWS CLI and then threading that through a variable that is set as the value for `desired_vcpu`.

File: aws/resource_aws_batch_compute_environment_test.go [hunk 1]
```diff
@@ -447,7 +473,8 @@ resource "aws_iam_role_policy_attachment" "aws_ec2_spot_fleet_role" {
 ########## security group ##########
 
 resource "aws_security_group" "test_acc" {
-  name = "tf_acc_test_batch_sg_%d"
+  name   = "tf_acc_test_batch_sg_%d"
+  vpc_id = "${aws_vpc.test_acc.id}"
```
Comment (hectcastro) [inline aws/resource_aws_batch_compute_environment_test.go]: This was added to help produce a valid CE when `desired_vcpu` is non-zero. If I made code changes to the resource such that it wouldn't send a value does desired vCPUs, the CE would get into an invalid state citing that its associated security group wasn't in the specified VPC.

**Comment by ratanos** [comment]
@hectcastro Thanks. I was planning to file a bug and came across this issue.
My terraform configuration snippet
```
min_vcpus = 2
desired_vcpus = 2
max_vcpus = 8
```
```
lifecycle {
ignore_changes["desired_vcpus"]
}
```
Even if AWS Batch does increase desired_vcpus, 
 _terraform apply_  should ignore desired_vcpus.

For some reason,
Terraform doesn't ignore "desired_vcpus".

Any idea if this is the expected behaviour or related the issue you're fixing ?

**Comment by euclideansphere** [comment]
seems like the value needs to be excluded from the api request (or included if we pass back the value retrieved from aws). any update on this? anything I can do to move this along?

**Comment by rajbettaswamy** [comment]
Is there any current workaround for this issue ?

**Comment by theothertom** [comment]
> Is there any current workaround for this issue ?
I do something like this:
 -var "batch_desired_cpu_workaround=$(aws batch describe-compute-environments --region eu-west-2 | jq '.computeEnvironments[0].computeResources.desiredvCpus')"
and then pass the desired CPU count to my batch compute environment

**Comment by rajbettaswamy** [comment]
Thanks for the workaround @theothertom. This solves my issue. But I had to modify the implementation to fit my requirement for couple of reasons, like:
1. I use TF in CI/CD pipeline.
2. We have multiple AWS Batch compute environments, so I can't really query by computeEnvironments[0,1,2...n] using jq.
3. Also, all my AWS Batch compute environments name ends with random strings due to following issue #3207 . So I can't even query my compute environment by name.

So I used `external ` data source template in Terraform:

```
data "external" "aws_batch_workaround" {
  program = ["bash", "${path.module}/scripts/desired_vcpus.sh"]

  query = {
    environment = "${var.environment}"
    name        = "${var.config["name"]}"
  }
}
```

Script:

```
cat desired_vcpus.sh

#!/bin/bash

function check_dependency() {
  test -f $(which jq) || error_exit "jq command not be detected in the path, please install it"
}

function parse_input() {
  eval "$(jq -r '@sh "export ENVIRONMENT=\(.environment) NAME=\(.name)"')"
  if [[ -z "${ENVIRONMENT}" ]]; then export ENVIRONMENT=none; fi
  if [[ -z "${NAME}" ]]; then export NAME=none; fi
}

function fetch_desired_vcpus() {
  desired_vcpus_contents=`aws batch describe-compute-environments --compute-environments $(aws batch describe-compute-environments | grep 'computeEnvironmentName' | awk '{print $2}' | grep '${ENVIRONMENT}-${NAME}.*' | cut -c 2- | rev | cut -c 3- | rev) | jq '.computeEnvironments[0].computeResources.desiredvCpus'`
}

function output() {
  jq -n \
    --arg desired_vcpus "$desired_vcpus_contents" \
    '{"desired_vcpus":$desired_vcpus}'
}

# main ()
check_dependency
parse_input
fetch_desired_vcpus
output
```

And referenced like this:
```
resource "aws_batch_compute_environment" "batch" {
  ...
  compute_resources {
  ...
    desired_vcpus      = "${data.external.aws_batch_workaround.result.desired_vcpus}"
  }
}
```
Inspired from this [gist](https://gist.github.com/irvingpop/968464132ded25a206ced835d50afa6b)

**Comment by ratanos** [comment]
@rajbettaswamy This one ignores desired_vcpus,
resource "aws_batch_compute_environment" "batch" {
lifecycle {
    ignore_changes = ["compute_resources.0.desired_vcpus"]
  }
}

**Comment by rajbettaswamy** [comment]
@ratanos Wow ! This solves my issue and implementation is way cleaner and simple compared to mine. Thanks for your help.

**Comment by kthompson** [comment]
Using the lifecycle ignore changes no longer appears to work in 0.12 of terraform 

ie
```terraform
  lifecycle {
    ignore_changes = [
      "compute_resources.0.desired_vcpus"
    ]
  }
```

**Comment by ratanos** [comment]
@hectcastro @bflad Can we merge this PR?

**Comment by chasezimmy** [comment]
If anyone is on Terraform 0.12 and looking for a solution => https://github.com/hashicorp/terraform/issues/21433#issuecomment-495829707

**Comment by nediml** [comment]
> If anyone is on Terraform 0.12 and looking for a solution => [hashicorp/terraform#21433 (comment)](https://github.com/hashicorp/terraform/issues/21433#issuecomment-495829707)
> 
> [ignore_changes docs](https://www.terraform.io/docs/configuration/resources.html#ignore_changes)

This only solves the issue when compute env is scaling down (not in all cases too).
When scaling up happens it is needed to set new value for desired vCPUs which has to be equal or higher than minimum vCPUs, and if we ignore desired vCPUs value, 0 will be used since it is default value (which is lower than current minimum vCPUs value).

Therefore, we still need this PR merged ASAP.

**Pull request merged**